File: content/problems/0139-word-break/solutions/prefix-dp/Solution.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> words = new HashSet<>(wordDict);
        boolean[] possible = new boolean[s.length() + 1];
        possible[0] = true;
        for (int end = 1; end <= s.length(); end++) {
            for (int start = 0; start < end; start++) {
                if (possible[start] && words.contains(s.substring(start, end))) {
                    possible[end] = true;
                    break;
                }
            }
        }
        return possible[s.length()];
    }
}

File: content/problems/0139-word-break/solutions/prefix-dp/Solution.kt
class Solution {
    fun wordBreak(s: String, wordDict: List<String>): Boolean {
        val words = wordDict.toHashSet()
        val possible = BooleanArray(s.length + 1)
        possible[0] = true
        for (end in 1..s.length) {
            for (start in 0 until end) {
                if (possible[start] && s.substring(start, end) in words) {
                    possible[end] = true
                    break
                }
            }
        }
        return possible[s.length]
    }
}
